### ISSUE TITLE
Kubevirt on ARM64

### ISSUE THREAD

**Issue by zhlhahaha** [kubevirt/kubevirt#3558]
**What you expected to happen**:
Enable **Kubevirt** and **kubevirt CI** on ARM64

Kubevirt on ARM64
-

- [ ] Check if all binaries and images in WORKSPACE have alternatives for ARM64 platform
--Three images do not support ARM64 version yet. kubevirt-testing, nfs-server and microlivecd_image.
 
- [x] Enable basic feature of kubevirt on ARM64 platform, which include: https://github.com/kubevirt/kubevirt/pull/3415

  - General Make command works on ARM64, e.g. 
make, make push, make go-build, make bazel-build ...

  - VM life cycle control via kubevirt on ARM64, 
Create a predefined VM
Schedule a VM on a Kubernetes cluster
Launch a VM
Stop a VM
Delete a VM

- [ ] Enable kubevirt CDI on ARM64 platform 

- [ ] additional features verification on ARM64, e.g.
SRIOV
VM migration
Multus
  

- [ ] enable divice plugin on ARM64

- [ ] enable sidecar on ARM64

crossbuild kubevirt
-
1. - [x] enable C compiler toolchain for cross-build  (https://github.com/zhlhahaha/bazel_aarch64_toolchain and https://github.com/zhlhahaha/bazel_aarch64_toolchain/tree/crosstool)
2. - [ ] add crossbuild toolchain in kubevirt  (Current Working on it)
3. - [ ] crossbuild works on x86 by running `make bazel-build` and `make bazel-build-images`

kubevirt CI on ARM64
-
Enable kubevirt test-infra on ARM64

1. - [ ] build multi-arch prow utility images for ARM64 kubevirt/project-infra#587 and kubevirt/project-infra#612
2. - [ ] build multi-arch bootstrap images
3. - [x] integrate one bare metal arm64 server into test-infra.

- [x] make cluster-up via kind works on ARM64 platform (as arm64 server not support nested virtualization, we can use kind to run test.)
-- kind node image need to be built on ARM64, as there is no official release. This may be solved in 2021, please refer to https://github.com/kubernetes-sigs/kind/issues/166

 - [ ] make bazel-test all pass on ARM64 platform

- [ ] make functest all pass on ARM64 platform

**Comment by zhlhahaha** [comment]
cc @lubinszARM

**Comment by zhlhahaha** [comment]
Hi, @rmohr @vladikr @davidvossel
I have writen a task proposal for kubevirts on ARM64. Are there anything I miss, or do you have any suggestion on it?

**Comment by rmohr** [comment]
@zhlhahaha yes looks good. Sounds like you have everything there which is needed.

**Comment by vielmetti** [comment]
Looking forward to hearing progress on this, as https://github.com/rancher/harvester/issues/371 (Harvester., a new hyperconverged infrastrucure from Rancher) depends on this working.

**Comment by rmohr** [comment]
> Looking forward to hearing progress on this, as [rancher/harvester#371](https://github.com/rancher/harvester/issues/371) (Harvester., a new hyperconverged infrastrucure from Rancher) depends on this working.

@vielmetti our main issue right now for supporting this is actually having ARM64 machines for building and CI. If rancher depends on it, maybe you can start a thread on the kubevirt-dev mailing list and see what we can do  about this?

**Comment by vielmetti** [comment]
Thanks @rmohr - I put something on the community meeting agenda for Wednesday - I hope to have a brief writeup of the current situation to share before then.

**Comment by zhlhahaha** [comment]
Hi @vielmetti @rmohr, here is a summary of current status for kubevirt on arm64.
1. The main PR https://github.com/kubevirt/kubevirt/pull/3415 has updated last week and works fine on ARM64 platform.
2. Prow image builder task need to be rebased. Here are two main task for it. I will fix them this week.  
- - https://github.com/kubevirt/project-infra/pull/587
- - https://github.com/kubevirt/project-infra/pull/612
3. The PR for building mulitarch test image of kubevirt is blocked as one reviewer does not response.
- - https://github.com/kubevirt/kubevirtci/pull/406

**Comment by zhlhahaha** [comment]
Hi @rmohr @davidvossel I have update the todo list at the begin of this Issue and add some tasks for crossbuild kubevirt.

Updated summary of current status for kubevirt on arm64.
1. With great support from @rmohr , bazeldnf works fine on ARM64 and `make rpm-deps` works well to auto-generate rpms informations.
2. Investigation on cross-build kubevirt for ARM64 on x86_64 server. And now blocked on the following bazel failures https://github.com/bazelbuild/bazel/issues/12859

**Comment by advancedwebdeveloper** [comment]
> ~/kubevirt$ make
> hack/dockerized "DOCKER_PREFIX= DOCKER_TAG= IMAGE_PULL_POLICY= VERBOSITY= ./hack/build-manifests.sh && \
>     hack/bazel-fmt.sh && hack/bazel-build.sh"
> hack/dockerized "DOCKER_PREFIX= DOCKER_TAG= IMAGE_PULL_POLICY= VERBOSITY= ./hack/build-manifests.sh && \
>     hack/bazel-fmt.sh && hack/bazel-build.sh"
> selecting docker as container runtime
> kubevirt
> Unable to find image 'quay.io/kubevirt/builder:2103171450-7edc9308c' locally
> 2103171450-7edc9308c: Pulling from kubevirt/builder
> docker: no matching manifest for linux/arm64/v8 in the manifest list entries.
> See 'docker run --help'.
> make: *** [Makefile:9: all] Error 125

**Comment by zhlhahaha** [comment]
> > ~/kubevirt$ make
> > hack/dockerized "DOCKER_PREFIX= DOCKER_TAG= IMAGE_PULL_POLICY= VERBOSITY= ./hack/build-manifests.sh && 
> > hack/bazel-fmt.sh && hack/bazel-build.sh"
> > hack/dockerized "DOCKER_PREFIX= DOCKER_TAG= IMAGE_PULL_POLICY= VERBOSITY= ./hack/build-manifests.sh && 
> > hack/bazel-fmt.sh && hack/bazel-build.sh"
> > selecting docker as container runtime
> > kubevirt
> > Unable to find image 'quay.io/kubevirt/builder:2103171450-7edc9308c' locally
> > 2103171450-7edc9308c: Pulling from kubevirt/builder
> > docker: no matching manifest for linux/arm64/v8 in the manifest list entries.
> > See 'docker run --help'.
> > make: *** [Makefile:9: all] Error 125

Please use this PR. https://github.com/kubevirt/kubevirt/pull/3415

**Comment by vielmetti** [comment]
Looks like #3415 was merged - however the latest build at https://github.com/kubevirt/kubevirt/releases/tag/v0.42.1 does not have an arm64 version of `virtctl`.

**Comment by rmohr** [comment]
The code is now ready for arm but the e2e testing is still woo. Before we can at least provide a minimum guarantee that arm works (running a critical mass of e2e tests) we will not release arm artifacts. You can cross compile kubevirt at this stage easily.

Having said that, a virtctl version of virtctl could be provided any time. It has nothing to do with arm support in general. It would fall into the same category as our windows virtctl.

Is that answer helpful?

**Comment by anyushun** [comment]
Hello @zhlhahaha, I found @luoyunhe working on: `Enable kubevirt CDI on ARM64 platform`, commit: https://github.com/kubevirt/containerized-data-importer/commit/3cfcd4f26137a4554bc4a3834100ba0b5bdceb55

**Comment by zhlhahaha** [comment]
> Hello @zhlhahaha, I found @luoyunhe working on: `Enable kubevirt CDI on ARM64 platform`, commit: [kubevirt/containerized-data-importer@3cfcd4f](https://github.com/kubevirt/containerized-data-importer/commit/3cfcd4f26137a4554bc4a3834100ba0b5bdceb55)

Why can not I find this PR in kubevirt/containerized-data-importer project? But I can see this commit in @luoyunhe 's own fork.

**Comment by rmohr** [comment]
> Why can not I find this PR in kubevirt/containerized-data-importer project? But I can see this commit in @luoyunhe 's own fork.

Seems to be a github feature to reference commits this way on forks. The top of the pages says

```
This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.
```

**Comment by zhlhahaha** [comment]
> > Why can not I find this PR in kubevirt/containerized-data-importer project? But I can see this commit in @luoyunhe 's own fork.
> 
> Seems to be a github feature to reference commits this way on forks. The top of the pages says
> 
> ```
> This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.
> ```

Yes, I can see the PR in @luoyunhe 's fork. https://github.com/luoyunhe/containerized-data-importer/tree/v1.34.1-aarch64

**Comment by zhlhahaha** [comment]
Hi, @luoyunhe, would you like to submit your PR which enables building multi-arch CDI? It would be helpful. Thanks.

**Comment by luoyunhe** [comment]
> Hi, @luoyunhe, would you like to submit your PR which enables building multi-arch CDI? It would be helpful. Thanks.

hi, I modified it with reference to your code, but there are still some problems. E.g. test and vddk.

**Comment by zhlhahaha** [comment]
> vddk

NP, if you meet some problems that hard to solve and need help, you can ping me.

**Comment by luoyunhe** [comment]
> > hi, I modified it with reference to your code, but there are still some problems. E.g. test and vddk.
> 
> NP, if you meet some problems that hard to solve and need help, you can ping me.

OK，thanks

**Comment by luoyunhe** [comment]
Hi, @zhlhahaha, does the aarch64 architecture support vddk data source? shoud i disable vddk feature in aarch64.

**Comment by zhlhahaha** [comment]
> Hi, @zhlhahaha, does the aarch64 architecture support vddk data source? shoud i disable vddk feature in aarch64.

I never use VDDK before, and there is only x86_64 version VDDK on the official website, so I think you can disable this feature for now.

**Comment by zhlhahaha** [comment]
> Hi, @zhlhahaha, does the aarch64 architecture support vddk data source? shoud i disable vddk feature in aarch64.

I post an issue here to see if VDDK supports ARM64.
https://github.com/vmware/virtual-disks/issues/9

**Comment by kwiesmueller** [comment]
Great to see this!
One question/potential task would be if we can get multi-arch images for this.
We tried deploying kubevirt with a custom ARM build on a cluster made from mixed nodes (Raspberry Pi and regular VMs) and right now it's not possible to do.
To support mixed architectures right now we need to:
a) build multi-arch images
b) use tags for pulling images (as the internal SHAs will be different)
I didn't manage to do this yet, partially because the Bazel rules don't support multi-arch images. We might need a separate build process merging two images somehow into a multi-arch one.

**Comment by zhlhahaha** [comment]
> We tried deploying kubevirt with a custom ARM build on a cluster made from mixed nodes (Raspberry Pi and regular VMs) and right now it's not possible to do.

Now we have experimental release for ARM64, and here is the guide for how to deploy kubevirt on ARM64 https://kubevirt.io/user-guide/operations/installation/#experimental-arm64-developer-builds
I have not tried to deploy it on Raspberry Pi, so please tell me if you meet any problems.

By the way, the arm64 vmi images are also in `quay.io/kubevirt`,  for example:
```
LATEST=$(curl -L https://storage.googleapis.com/kubevirt-prow/devel/nightly/release/kubevirt/kubevirt/latest-arm64)
docker pull quay.io/kubevirt/fedora-cloud-container-disk-demo:${LATEST}-arm64
```

> s don't support multi-arch images. We might need a separate build process merging two images somehow into a multi-arch one.

As images for ARM64 arch is experimental and many features have not been verified or implied, we are not supposed to release a multi-arch one until kubeVirt on ARM64 is as mature as that on x86_64.

**Comment by rmohr** [comment]
> a) build multi-arch images
> b) use tags for pulling images (as the internal SHAs will be different)
> I didn't manage to do this yet, partially because the Bazel rules don't support multi-arch images. We might need a separate build process merging two images somehow into a multi-arch one.

We can simply push all images and then afterwards build a multi-arch image with e.g. `dockerx`. That should not be too hard. We did that in a few other contexts.

However: virt-handler and virt-api have no understanding what mixed cluster means. It just knows about the architecture it is running on. Supporting mixed clusters is in my opinion its own epic outside of supporting arm.

**Comment by colinblack** [comment]
Hi, @zhlhahaha, where can i get virtctl for Arm64.

**Comment by zhlhahaha** [comment]
> Hi, @zhlhahaha, where can i get virtctl for Arm64.

As we only compile amd64 virtctl in the upstream now, you need to compile Arm64 virtctl by yourself.
Here are two ways 
for native build:
```
bazel build //cmd/virtctl:virtctl
```
for cross build:
```
bazel build --config=crossbuild-aarch64 //cmd/virtctl:virtctl
```

**Comment by colinblack** [comment]
> > Hi, @zhlhahaha, where can i get virtctl for Arm64.
> 
> As we only compile amd64 virtctl in the upstream now, you need to compile Arm64 virtctl by yourself.
> Here are two ways
> for native build:
> 
> ```
> bazel build //cmd/virtctl:virtctl
> ```
> 
> for cross build:
> 
> ```
> bazel build --config=crossbuild-aarch64 //cmd/virtctl:virtctl
> ```
OK, thanks

**Comment by luoyunhe** [comment]
@zhlhahaha I encountered some [problems](https://github.com/kubevirt/containerized-data-importer/pull/1863) during the pipeline test, can you provide me with some help?

**Comment by awels** [comment]
Note on the CDI side we had never considered aarch64 so it is highly likely our bazel configuration is not flexible enough, in particular our bazelrc. But we are very much open to supporting it, so if I can help in some way.

**Comment by zhlhahaha** [comment]
> Hi, @zhlhahaha I encountered some [problems](https://github.com/kubevirt/containerized-data-importer/pull/1863) during the pipeline test, can you provide me with some help? thanks.

I think @awels has already given you some comments. Let see how the CI test result. I will give it a try today or tomorrow, then give you a reply.

**Comment by luoyunhe** [comment]
> > Hi, @zhlhahaha I encountered some [problems](https://github.com/kubevirt/containerized-data-importer/pull/1863) during the pipeline test, can you provide me with some help? thanks.
> 
> I think @awels has already given you some comments. Let see how the CI test result. I will give it a try today or tomorrow, then give you a reply.

Ok, thanks.

**Comment by kbc8894** [comment]
I have a problem when I test kubevirt on arm64. how to handle it?
kubectl virt console is work.

kubectl virt vnc is not work.
```
Can't access VMI ubuntu-desktop-vm: No graphics devices are present.
```
so I try to vnc with [repo](https://github.com/wavezhang/virtVNC)
result is it.
<img width="1030" alt="Screen Shot 2021-08-29 at 5 44 15 PM" src="https://user-images.githubusercontent.com/3617939/131244438-f2312ca8-6291-4252-ad09-ad0c3bc70df6.png">



below is VirtualMachine spec.
ubuntu iso image is downloaded from [here](https://cdimage.ubuntu.com/focal/daily-live/current/).
```
---
 apiVersion: v1
 kind: PersistentVolumeClaim
 metadata:
   name: ubuntu-desktop-hd
 spec:
   accessModes:
     - ReadWriteOnce
   resources:
     requests:
       storage: 15Gi

---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: ubuntu-desktop-vm
spec:
  running: false
  template:
    metadata:
      labels:
        kubevirt.io/size: small
        kubevirt.io/domain: ubuntu-desktop-vm
    spec:
      domain:
        cpu:
          cores: 2
        devices:
          disks:
            - bootOrder: 1
              name: cdromiso
              disk:
                bus: virtio
                readonly: true
            - name: harddrive
              disk:
                bus: virtio
        machine:
          type: virt
        resources:
          requests:
            memory: 2G
      volumes:
        - name: cdromiso
          persistentVolumeClaim:
            claimName: ubuntu-desktop-iso
        - name: harddrive
          persistentVolumeClaim:
            claimName: ubuntu-desktop-hd
```

**Comment by zhlhahaha** [comment]
> ```
> Can't access VMI ubuntu-desktop-vm: No graphics devices are present.
> ```

Yes, graphics devices has disabled on ARM, please refer to the following code
https://github.com/kubevirt/kubevirt/blob/ef3e07525537d7b3ee1fa4c7d0d524bbd6eca93f/pkg/virt-api/webhooks/arm64.go#L65-L71

The root cause for this is that the qemu-kvm binary in virt-launcher does not support vga devices. For more detail, you can read the discuss here https://github.com/kubevirt/libvirt/issues/49

**Comment by kbc8894** [comment]
> > ```
> > Can't access VMI ubuntu-desktop-vm: No graphics devices are present.
> > ```
> 
> 
> @kbc8894 Yes, graphics devices has disabled on ARM, please refer to the following code
> 
> 
> https://github.com/kubevirt/kubevirt/blob/ef3e07525537d7b3ee1fa4c7d0d524bbd6eca93f/pkg/virt-api/webhooks/arm64.go#L65-L71
> 
> 
> 
> The root cause for this is that the qemu-kvm binary in virt-launcher does not support vga devices. For more detail, you can read the discuss here [kubevirt/libvirt#49](https://github.com/kubevirt/libvirt/issues/49)

@zhlhahaha Thanks for your explanation!

**Comment by rmohr** [comment]
On Wed, Sep 1, 2021 at 5:08 AM zhlhahaha ***@***.***> wrote:

> Can't access VMI ubuntu-desktop-vm: No graphics devices are present.
>
> Yes, graphics devices has disabled on ARM, please refer to the following
> code
>
> https://github.com/kubevirt/kubevirt/blob/ef3e07525537d7b3ee1fa4c7d0d524bbd6eca93f/pkg/virt-api/webhooks/arm64.go#L65-L71
>
> The root cause for this is that the qemu-kvm binary in virt-launcher does
> not support vga devices. For more detail, you can read the discuss here
> kubevirt/libvirt#49 <https://github.com/kubevirt/libvirt/issues/49>
>

Would it make sense to look into virtio-gpu like suggested by Andrea and
use it instead for arm as the default device?
I guess in the worst case, older ARM guests would just not understand the
device and would still not have VNC, but newer ones would have it.

Best regards,
Roman

**Comment by wonleing** [comment]
Seems basic functions work already. Is it possible to release a tech-preview for arm64 (by disabling non-work parts)?
So we can carry on upper development (like harvester) at the same time.

**Comment by zhlhahaha** [comment]
> Would it make sense to look into virtio-gpu like suggested by Andrea and use it instead for arm as the default device? I guess in the worst case, older ARM guests would just not understand the device and would still not have VNC, but newer ones would have it. Best regards, Roman

Sorry, I missed this comment. I have not verified virtio-gpu on arm64 server yet, but I think this should work, I will give it a try.

**Comment by zhlhahaha** [comment]
> Seems basic functions work already. Is it possible to release a tech-preview code branch or tag for arm64 (by disabling non-work parts)?

Hi Leon, I am not sure how this works, as there are no clarified parts that are not supported by ARM64, but we do have some parts that are not verified yet, like SRIOV and Multus. Roman @rmohr, do you have any comments?

> PS: Is the TODO check list on the top up-to-date? We want to make some contribution if there's any key gap (except CI and test stuff).

Yes, it is keeping updated, currently, I am doing CDI parts. The key gap is that many kubevirt features have not been verified on ARM64 server, like SRIOV, Multus, CPU pinning, and so forth. It would be great if you can list any features that you need but are not on the list.
Welcome to do any contributions!!!

**Comment by rmohr** [comment]
> Seems basic functions work already. Is it possible to release a tech-preview code branch or tag for arm64 (by disabling non-work parts)?

As @zhlhahaha said, there should only be specific features which are not working which we can enable step-by-step. Identifying and disabling specific parts would require us to maintain and test these switches too. I would rather prefer to increase the CI coverage over time and fix specific features along the way and securing the progress with passing CI tests.

> So we can carry on upper development (like harvester) at the same time.

I think you can just start working with the nightly developer releases: https://kubevirt.io/user-guide/operations/installation/#experimental-arm64-developer-builds.

Let us know if you think that this does not work for you.

**Comment by zhlhahaha** [comment]
> @zhlhahaha We're having trouble with the demo for KubeVirt on ARM. Do you have any time this week to jump on a zoom call to assist with troubleshooting? Thanks!

Ok, I will contact you on slack.

**Comment by hellocloudnative** [comment]
@zhlhahaha

**Comment by mazzystr** [comment]
I'm making some forward steps.  I pulled one rpi out of the cluster and installed minikube k8s with KubeVirt.  KubeVirt fully deploys.  I am now able to launch a virtual machine.

> \# kubectl get all -A
> NAMESPACE     NAME                                  READY   STATUS    RESTARTS      AGE
> default       pod/virt-launcher-testvm-9zh5v        2/2     Running   0             18m
> kube-system   pod/coredns-78fcd69978-5gbjp          1/1     Running   0             65m
> kube-system   pod/etcd-gold1                        1/1     Running   0             66m
> kube-system   pod/kube-apiserver-gold1              1/1     Running   0             66m
> kube-system   pod/kube-controller-manager-gold1     1/1     Running   0             66m
> kube-system   pod/kube-proxy-t9gns                  1/1     Running   0             65m
> kube-system   pod/kube-scheduler-gold1              1/1     Running   0             66m
> kube-system   pod/storage-provisioner               1/1     Running   1 (45m ago)   65m
> kubevirt      pod/virt-api-db7f949b4-5vx9w          1/1     Running   0             53m
> kubevirt      pod/virt-api-db7f949b4-l72kw          1/1     Running   0             53m
> kubevirt      pod/virt-controller-7b7d99ff5-9m7mt   1/1     Running   0             52m
> kubevirt      pod/virt-controller-7b7d99ff5-hpxcd   1/1     Running   1 (45m ago)   52m
> kubevirt      pod/virt-handler-x6sl4                1/1     Running   0             52m
> kubevirt      pod/virt-operator-55bbc8984f-5dssj    1/1     Running   0             58m
> kubevirt      pod/virt-operator-55bbc8984f-zqkhq    1/1     Running   1 (45m ago)   58m
> 
> NAMESPACE     NAME                                  TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                  AGE
> default       service/kubernetes                    ClusterIP   10.96.0.1        <none>        443/TCP                  66m
> kube-system   service/kube-dns                      ClusterIP   10.96.0.10       <none>        53/UDP,53/TCP,9153/TCP   66m
> kubevirt      service/kubevirt-operator-webhook     ClusterIP   10.109.138.249   <none>        443/TCP                  53m
> kubevirt      service/kubevirt-prometheus-metrics   ClusterIP   10.97.246.93     <none>        443/TCP                  53m
> kubevirt      service/virt-api                      ClusterIP   10.109.169.118   <none>        443/TCP                  53m
> 
> NAMESPACE     NAME                          DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
> kube-system   daemonset.apps/kube-proxy     1         1         1       1            1           kubernetes.io/os=linux   66m
> kubevirt      daemonset.apps/virt-handler   1         1         1       1            1           kubernetes.io/os=linux   52m
> 
> NAMESPACE     NAME                              READY   UP-TO-DATE   AVAILABLE   AGE
> kube-system   deployment.apps/coredns           1/1     1            1           66m
> kubevirt      deployment.apps/virt-api          2/2     2            2           53m
> kubevirt      deployment.apps/virt-controller   2/2     2            2           52m
> kubevirt      deployment.apps/virt-operator     2/2     2            2           58m
> 
> NAMESPACE     NAME                                        DESIRED   CURRENT   READY   AGE
> kube-system   replicaset.apps/coredns-78fcd69978          1         1         1       65m
> kubevirt      replicaset.apps/virt-api-db7f949b4          2         2         2       53m
> kubevirt      replicaset.apps/virt-controller-7b7d99ff5   2         2         2       52m
> kubevirt      replicaset.apps/virt-operator-55bbc8984f    2         2         2       58m
> 
> NAMESPACE   NAME                                AGE   STATUS    READY
> default     virtualmachine.kubevirt.io/testvm   18m   Running   True
> 
> NAMESPACE   NAME                                        AGE   PHASE     IP            NODENAME   READY
> default     virtualmachineinstance.kubevirt.io/testvm   18m   Running   172.17.0.10   gold1      True
> 
> NAMESPACE   NAME                            AGE   PHASE
> kubevirt    kubevirt.kubevirt.io/kubevirt   53m   Deployed


Next problems
* I am unable to ping the vm IP yet I can ping the docker hub ip of 172.17.0.1
* I need to compile virtctl for arm.  See https://github.com/kubevirt/kubevirt/issues/3558

**Comment by mazzystr** [comment]
virtctl compiles successfully

Console on the virtual machine is empty.

```
# kc describe virtualmachine.kubevirt.io/testvm
Name:         testvm
Namespace:    default
Labels:       <none>
Annotations:  kubevirt.io/latest-observed-api-version: v1
              kubevirt.io/storage-observed-api-version: v1alpha3
API Version:  kubevirt.io/v1
Kind:         VirtualMachine
Metadata:
  Creation Timestamp:  2021-10-13T19:47:48Z
  Generation:          2
  Managed Fields:
    API Version:  kubevirt.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:template:
          .:
          f:metadata:
            .:
            f:labels:
          f:spec:
            .:
            f:domain:
              .:
              f:devices:
                .:
                f:disks:
                f:interfaces:
              f:resources:
                .:
                f:requests:
                  .:
                  f:memory:
            f:networks:
            f:volumes:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2021-10-13T19:47:48Z
    API Version:  kubevirt.io/v1alpha3
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:kubevirt.io/latest-observed-api-version:
          f:kubevirt.io/storage-observed-api-version:
      f:spec:
        f:running:
      f:status:
        .:
        f:volumeSnapshotStatuses:
    Manager:      Go-http-client
    Operation:    Update
    Time:         2021-10-13T19:48:02Z
    API Version:  kubevirt.io/v1alpha3
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:conditions:
        f:created:
        f:printableStatus:
        f:ready:
    Manager:         Go-http-client
    Operation:       Update
    Subresource:     status
    Time:            2021-10-13T19:53:01Z
  Resource Version:  25063
  UID:               501ce15d-5cb4-4935-b01f-d79c99ff0e14
Spec:
  Running:  true
  Template:
    Metadata:
      Creation Timestamp:  <nil>
      Labels:
        kubevirt.io/domain:  testvm
        kubevirt.io/size:    small
    Spec:
      Domain:
        Devices:
          Disks:
            Disk:
              Bus:  virtio
            Name:   containerdisk
            Disk:
              Bus:  virtio
            Name:   cloudinitdisk
          Interfaces:
            Masquerade:
            Name:  default
        Machine:
          Type:  virt
        Resources:
          Requests:
            Memory:  64M
      Networks:
        Name:  default
        Pod:
      Volumes:
        Container Disk:
          Image:  quay.io/kubevirt/alpine-container-disk-demo:20211005_139f486ec-arm64
        Name:     containerdisk
        Cloud Init No Cloud:
          userDataBase64:  SGkuXG4=
        Name:              cloudinitdisk
Status:
  Conditions:
    Last Probe Time:       <nil>
    Last Transition Time:  2021-10-13T19:52:56Z
    Status:                True
    Type:                  Ready
  Created:                 true
  Printable Status:        Running
  Ready:                   true
  Volume Snapshot Statuses:
    Enabled:  false
    Name:     containerdisk
    Reason:   Snapshot is not supported for this volumeSource type [containerdisk]
    Enabled:  false
    Name:     cloudinitdisk
    Reason:   Snapshot is not supported for this volumeSource type [cloudinitdisk]
Events:
  Type    Reason            Age                From                       Message
  ----    ------            ----               ----                       -------
  Normal  SuccessfulDelete  18m                virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 9b16760e-fb30-4772-93f1-ddb6740ec10a
  Normal  SuccessfulCreate  18m (x2 over 22m)  virtualmachine-controller  Started the virtual machine by creating the new virtual machine instance testvm
```

```
# kc describe virtualmachineinstance.kubevirt.io/testvm
Name:         testvm
Namespace:    default
Labels:       kubevirt.io/domain=testvm
              kubevirt.io/nodeName=gold1
              kubevirt.io/size=small
Annotations:  kubevirt.io/latest-observed-api-version: v1
              kubevirt.io/storage-observed-api-version: v1alpha3
API Version:  kubevirt.io/v1
Kind:         VirtualMachineInstance
Metadata:
  Creation Timestamp:  2021-10-13T19:52:46Z
  Finalizers:
    kubevirt.io/virtualMachineControllerFinalize
    foregroundDeleteVirtualMachine
  Generation:  8
  Managed Fields:
    API Version:  kubevirt.io/v1alpha3
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubevirt.io/latest-observed-api-version:
          f:kubevirt.io/storage-observed-api-version:
        f:finalizers:
          .:
          v:"kubevirt.io/virtualMachineControllerFinalize":
        f:labels:
          .:
          f:kubevirt.io/domain:
          f:kubevirt.io/nodeName:
          f:kubevirt.io/size:
        f:ownerReferences:
          .:
          k:{"uid":"501ce15d-5cb4-4935-b01f-d79c99ff0e14"}:
      f:spec:
        .:
        f:domain:
          .:
          f:devices:
            .:
            f:disks:
            f:interfaces:
          f:firmware:
            .:
            f:uuid:
          f:machine:
            .:
            f:type:
          f:resources:
            .:
            f:requests:
              .:
              f:memory:
        f:networks:
        f:volumes:
      f:status:
        .:
        f:activePods:
          .:
          f:a34864dd-9c93-4c9d-a262-00ae78db0fdc:
        f:conditions:
        f:guestOSInfo:
        f:interfaces:
        f:launcherContainerImageVersion:
        f:migrationMethod:
        f:migrationTransport:
        f:nodeName:
        f:phase:
        f:phaseTransitionTimestamps:
        f:qosClass:
        f:virtualMachineRevisionName:
        f:volumeStatus:
    Manager:    Go-http-client
    Operation:  Update
    Time:       2021-10-13T19:53:01Z
  Owner References:
    API Version:           kubevirt.io/v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  VirtualMachine
    Name:                  testvm
    UID:                   501ce15d-5cb4-4935-b01f-d79c99ff0e14
  Resource Version:        25061
  UID:                     d30a7d14-9145-4c2a-bb51-1a6e3ba8a8f4
Spec:
  Domain:
    Cpu:
      Cores:    1
      Model:    host-passthrough
      Sockets:  1
      Threads:  1
    Devices:
      Autoattach Graphics Device:  false
      Disks:
        Disk:
          Bus:  virtio
        Name:   containerdisk
        Disk:
          Bus:  virtio
        Name:   cloudinitdisk
      Interfaces:
        Masquerade:
        Name:  default
    Features:
      Acpi:
        Enabled:  true
    Firmware:
      Bootloader:
        Efi:
          Secure Boot:  false
      Uuid:             5a9fc181-957e-5c32-9e5a-2de5e9673531
    Machine:
      Type:  virt
    Resources:
      Requests:
        Memory:  64M
  Networks:
    Name:  default
    Pod:
  Volumes:
    Container Disk:
      Image:              quay.io/kubevirt/alpine-container-disk-demo:20211005_139f486ec-arm64
      Image Pull Policy:  IfNotPresent
    Name:                 containerdisk
    Cloud Init No Cloud:
      userDataBase64:  SGkuXG4=
    Name:              cloudinitdisk
Status:
  Active Pods:
    a34864dd-9c93-4c9d-a262-00ae78db0fdc:  gold1
  Conditions:
    Last Probe Time:       <nil>
    Last Transition Time:  2021-10-13T19:52:56Z
    Status:                True
    Type:                  Ready
    Last Probe Time:       <nil>
    Last Transition Time:  <nil>
    Status:                True
    Type:                  LiveMigratable
  Guest OS Info:
  Interfaces:
    Ip Address:  172.17.0.10
    Ip Addresses:
      172.17.0.10
    Mac:                             52:54:00:d9:5f:fb
    Name:                            default
  Launcher Container Image Version:  quay.io/kubevirt/virt-launcher:20211006_14d80cc1f-arm64
  Migration Method:                  BlockMigration
  Migration Transport:               Unix
  Node Name:                         gold1
  Phase:                             Running
  Phase Transition Timestamps:
    Phase:                        Pending
    Phase Transition Timestamp:   2021-10-13T19:52:46Z
    Phase:                        Scheduling
    Phase Transition Timestamp:   2021-10-13T19:52:47Z
    Phase:                        Scheduled
    Phase Transition Timestamp:   2021-10-13T19:52:56Z
    Phase:                        Running
    Phase Transition Timestamp:   2021-10-13T19:53:01Z
  Qos Class:                      Burstable
  Virtual Machine Revision Name:  revision-start-vm-501ce15d-5cb4-4935-b01f-d79c99ff0e14-2
  Volume Status:
    Name:    cloudinitdisk
    Target:  vdb
    Name:    containerdisk
    Target:  vda
Events:
  Type    Reason            Age                  From                       Message
  ----    ------            ----                 ----                       -------
  Normal  SuccessfulCreate  18m                  virtualmachine-controller  Created virtual machine pod virt-launcher-testvm-smq9h
  Normal  Started           18m                  virt-handler               VirtualMachineInstance started.
  Normal  Created           2m26s (x7 over 18m)  virt-handler               VirtualMachineInstance defined.
```

```
uuidd      29109   28883 99 19:53 ?        00:19:55 /usr/libexec/qemu-kvm -name guest=default_testvm,debug-threads=on -S -object secret,id=masterKey0,format=raw,file=/var/lib/libvirt/qemu/domain-1-default_testvm/master-key.aes -blockdev {"driver":"file","filename":"/usr/share/AAVMF/AAVMF_CODE.fd","node-name":"libvirt-pflash0-storage","auto-read-only":true,"discard":"unmap"} -blockdev {"node-name":"libvirt-pflash0-format","read-only":true,"driver":"raw","file":"libvirt-pflash0-storage"} -blockdev {"driver":"file","filename":"/tmp/default_testvm","node-name":"libvirt-pflash1-storage","auto-read-only":true,"discard":"unmap"} -blockdev {"node-name":"libvirt-pflash1-format","read-only":false,"driver":"raw","file":"libvirt-pflash1-storage"} -machine virt-rhel8.4.0,accel=kvm,usb=off,dump-guest-core=off,gic-version=2,pflash0=libvirt-pflash0-format,pflash1=libvirt-pflash1-format,memory-backend=mach-virt.ram -cpu host -m 62 -object memory-backend-ram,id=mach-virt.ram,size=65011712 -overcommit mem-lock=off -smp 1,sockets=1,dies=1,cores=1,threads=1 -object iothread,id=iothread1 -uuid 5a9fc181-957e-5c32-9e5a-2de5e9673531 -display none -no-user-config -nodefaults -chardev socket,id=charmonitor,fd=18,server,nowait -mon chardev=charmonitor,id=monitor,mode=control -rtc base=utc -no-shutdown -boot strict=on -device pcie-root-port,port=0x8,chassis=1,id=pci.1,bus=pcie.0,multifunction=on,addr=0x1 -device pcie-root-port,port=0x9,chassis=2,id=pci.2,bus=pcie.0,addr=0x1.0x1 -device pcie-root-port,port=0xa,chassis=3,id=pci.3,bus=pcie.0,addr=0x1.0x2 -device pcie-root-port,port=0xb,chassis=4,id=pci.4,bus=pcie.0,addr=0x1.0x3 -device pcie-root-port,port=0xc,chassis=5,id=pci.5,bus=pcie.0,addr=0x1.0x4 -device pcie-root-port,port=0xd,chassis=6,id=pci.6,bus=pcie.0,addr=0x1.0x5 -device pcie-root-port,port=0xe,chassis=7,id=pci.7,bus=pcie.0,addr=0x1.0x6 -device virtio-scsi-pci-non-transitional,id=scsi0,bus=pci.2,addr=0x0 -device virtio-serial-pci-non-transitional,id=virtio-serial0,bus=pci.3,addr=0x0 -blockdev {"driver":"file","filename":"/var/run/kubevirt/container-disks/disk_0.img","node-name":"libvirt-3-storage","cache":{"direct":true,"no-flush":false},"auto-read-only":true,"discard":"unmap"} -blockdev {"node-name":"libvirt-3-format","read-only":true,"discard":"unmap","cache":{"direct":true,"no-flush":false},"driver":"raw","file":"libvirt-3-storage"} -blockdev {"driver":"file","filename":"/var/run/kubevirt-ephemeral-disks/disk-data/containerdisk/disk.qcow2","node-name":"libvirt-2-storage","cache":{"direct":true,"no-flush":false},"auto-read-only":true,"discard":"unmap"} -blockdev {"node-name":"libvirt-2-format","read-only":false,"discard":"unmap","cache":{"direct":true,"no-flush":false},"driver":"qcow2","file":"libvirt-2-storage","backing":"libvirt-3-format"} -device virtio-blk-pci-non-transitional,bus=pci.4,addr=0x0,drive=libvirt-2-format,id=ua-containerdisk,bootindex=1,write-cache=on,werror=stop,rerror=stop -blockdev {"driver":"file","filename":"/var/run/kubevirt-ephemeral-disks/cloud-init-data/default/testvm/noCloud.iso","node-name":"libvirt-1-storage","cache":{"direct":true,"no-flush":false},"auto-read-only":true,"discard":"unmap"} -blockdev {"node-name":"libvirt-1-format","read-only":false,"discard":"unmap","cache":{"direct":true,"no-flush":false},"driver":"raw","file":"libvirt-1-storage"} -device virtio-blk-pci-non-transitional,bus=pci.5,addr=0x0,drive=libvirt-1-format,id=ua-cloudinitdisk,write-cache=on,werror=stop,rerror=stop -netdev tap,fd=20,id=hostua-default,vhost=on,vhostfd=21 -device virtio-net-pci-non-transitional,host_mtu=1500,netdev=hostua-default,id=ua-default,mac=52:54:00:d9:5f:fb,bus=pci.1,addr=0x0,romfile= -chardev socket,id=charserial0,fd=22,server,nowait -serial chardev:charserial0 -chardev socket,id=charchannel0,fd=23,server,nowait -device virtserialport,bus=virtio-serial0.0,nr=1,chardev=charchannel0,id=channel0,name=org.qemu.guest_agent.0 -device virtio-balloon-pci-non-transitional,id=balloon0,bus=pci.6,addr=0x0 -sandbox on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny -msg timestamp=on
```

**Comment by zhlhahaha** [comment]
> Does the qemu-KVM binary in the latest version of virt-Launcher support VGA devices? 

No, it does not support VGA devices, please refer to https://github.com/kubevirt/kubevirt/issues/3558#issuecomment-909835219

> If not, how can I replace qemu-KVM in virt-Launcher?

You can try to run virt-Launcher directly via `docker run`, then install the qemu-kvm directly form fedora official repository or compile it by yourself. After that, you can use `docker commit` to save the new virt-Launcher image.

**Comment by hellocloudnative** [comment]
I want to run VNC graphical interface on ARM, is there any way?  @zhlhahaha

**Comment by hellocloudnative** [comment]
Is virtio-GPU configured in the VIRTUAL machine's YAML file? @zhlhahaha

**Comment by hellocloudnative** [comment]
How do I configure a supported Virtio-GPU? @zhlhahaha

**Comment by zhlhahaha** [comment]
> How do I configure a supported Virtio-GPU? @zhlhahaha

I do not know, I don't do any investigation on this, but this code is for kubevirt setting vnc which may help.
https://github.com/kubevirt/kubevirt/blob/d64731bde56065a1c61cba01bcc248fa3a2f7b7f/pkg/virt-launcher/virtwrap/converter/converter.go#L1699-L1720

**Comment by rmohr** [comment]
> How do I configure a supported Virtio-GPU? @zhlhahaha

This is right now not possible. We normally hardcode the vga device in theVMs (except if you select the headless VM options). @zhlhahaha so far just ensured that the vga device never gets added. If virtio-gpu would be a suitable replacment for vga, I would assume that we simply est the virtio-gpu device instead of the vga one.

**Comment by zhlhahaha** [comment]
> > How do I configure a supported Virtio-GPU? @zhlhahaha
> 
> This is right now not possible. We normally hardcode the vga device in theVMs (except if you select the headless VM options). @zhlhahaha so far just ensured that the vga device never gets added. If virtio-gpu would be a suitable replacment for vga, I would assume that we simply est the virtio-gpu device instead of the vga one.

Hi Roman, I verified the virtio-gpu on bare-metal libvirt, it works fine. I will verify it on kubvirt.

```
...
    <controller type='pci' index='4' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='4' port='0xb'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x3'/>
    </controller>
...
    <graphics type='vnc' port='-1' autoport='yes'>
      <listen type='address'/>
    </graphics>
    <video>
      <model type='virtio' heads='1' primary='yes'/>
      <address type='pci' domain='0x0000' bus='0x04' slot='0x00' function='0x0'/>
    </video>
```

**Comment by hellocloudnative** [comment]
I changed the code model= VGA to virtio, which can be accessed through VNC, but the mouse and keyboard can not move.
@zhlhahaha

**Comment by andreabolognani** [comment]
> Yes, you need to add virtio-keyboard and virtio-mouse which are all supported in qemu-kvm

I believe USB keyboard plus USB tablet is the preferred configuration. Even on x86, a tablet is usually chosen over a mouse because it works better in practice for whatever reason.

**Comment by hellocloudnative** [comment]
I tried this configuration and got an error as follows:
`The request is invalid: spec.template.spec.domain.devices.inputs[1].type: Input device can have only tablet type.`
@zhlhahaha

**Comment by zhlhahaha** [comment]
> I tried this configuration and got an error as follows: `The request is invalid: spec.template.spec.domain.devices.inputs[1].type: Input device can have only tablet type.` @zhlhahaha

Have you applied my patch?

**Comment by hellocloudnative** [comment]
I modified the relevant code and it can be created normally, but the keyboard still does not respond
`
    devices:
      autoattachGraphicsDevice: true
      inputs:
      - type: tablet
        bus: virtio
        name: tablet1
      - type: keyboard
        bus: virtio
        name: keyboard1
`

**Comment by zhlhahaha** [comment]
> I modified the relevant code and configuration, it can be created normally, but the keyboard still does not respond

I open a new issue for this, we can discuss it there.
https://github.com/kubevirt/kubevirt/issues/6614

**Comment by zhlhahaha** [comment]
Hi @hellocloudnative
I have update the patch
https://github.com/zhlhahaha/kubevirt/commit/cfead387025912a30e4cbbcf71cad7b18c0e2a8e

Apply the patch and recompile virtctl, then give it a try. It works fine on my arm64 server.

**Comment by zhlhahaha** [comment]
> but when I first started the system, the keyboard didn't work at boot.

I do not understand this.

**Comment by hellocloudnative** [comment]
> > 但是当我第一次启动系统时，键盘在启动时性能。
> 
> 我不明白。

Is a boot boot boot, before entering the system installation interface, my keyboard is not available.

**Comment by hellocloudnative** [comment]
> > I skipped booting through the VNC console, and the mouse and keyboard are available on the system installation screen. But when I started up, the system boot interface, my keyboard can not be used. Another problem is that I can't add virtual disks. I use the following configuration virt-launcher will report an error:
> > ```
> >           disks:
> >           - bootOrder: 1
> >             disk:
> >               bus: virtio
> >             name: cdromiso
> >           - disk:
> >              bus: virtio
> >             name: harddrive
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > {"component":"virt-launcher","level":"error","msg":"Cannot find 'dmidecode' in path: No such file or directory","pos":"virExec:627","subcomponent":"libvirt","thread":"38","timestamp":"2021-10-15T06:49:23.114000Z"}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @zhlhahaha
> 
> This is a minor error, which would not lead to vm boot failure. Have you tried this on x86_64 server?

It works on x86.

**Comment by wonleing** [comment]
I guess you need add a USB keyboard, instead of a virtio tablet.
It is not yet able to virtio driver on the grub stage.

**Comment by hellocloudnative** [comment]
This modification is successful. Thank you very much.
```
          inputs:
          - type: tablet
            bus:  virtio
            name: tablet01
          - type: keyboard
            bus: usb
            name: keyboard01
```

**Comment by hellocloudnative** [comment]
Which video Device I can use for an Quemu Arm64 VM for Windows 10 on Arm? if I try the ramfb Device I only see "Press Enter for DVD Boot" but after that no more Setup Screen. 
@zhlhahaha

**Comment by zhlhahaha** [comment]
> I have a question, how to add ramfb device? @zhlhahaha

I have not used this device before, so I do general research on this. It seems that you can only add this device via `<qemu:commandline>` as the example in the following website https://wiki.archlinux.org/title/Intel_GVT-g#Enable_RAMFB_display_(optional). 

In kubevirt, you may try to use a sidecar-hook to modify the domain xml and add the ramfb related xml config.
https://github.com/kubevirt/kubevirt/issues/3745#issuecomment-664898554

This is only a possible way (may work or may not work), you need to give it a try.

**Comment by hellocloudnative** [comment]
use a sidecar-hook to modify the domain xml and add the ramfb related xml config.

**Comment by hellocloudnative** [comment]
I want to enable Win 10 virtual machine on KUbevirt on ARM.

**Comment by andreabolognani** [comment]
> I think the only concern regarding to UEFI is the big memory consumption just for the UEFI boot part. I don't know why so much memory is needed, but I only see it as an issue if one would want to create very small VMs with < 256Mb.  But probably worth exploring if there are significant differences.

I just tried running a CirrOS 0.5.2 aarch64 VM with 128 MiB of RAM and indeed it panics early during boot. I had more luck with 150 MiB, but it was not consistently successful. We've seen VMs becoming unreliable on very low memory in the past, even on x86_64, so I guess considering 256 MiB to be the minimum makes sense.

> I don't know if that is even ARM specific. I guess we need quite some RAM also on x86_64, but there uefi is only used in our test suite for fedora where we anyway have to add more RAM than that just for the OS.

I haven't tried either, but my guess would be that UEFI is simply more resource hungry than BIOS. See also https://github.com/cirros-dev/cirros/issues/53.

> But probably worth exploring if there are significant differences.

Booting an aarch64 image that's not prepared for UEFI is not trivial, as the firmware in that case is neither included in the disk image nor already available as part of the host OS. I think you might be able to sideload the u-boot binary, but I've never tried that; I've done direct kernel boot in the past and that works, but it usually requires moving around one or more additional files (kernel plus possibly initrd and dtb) together with the disk image. It gets really unwieldy really fast, so it's not surprising that all aarch64 cloud images are assuming UEFI :)

> I want to enable Win 10 virtual machine on KUbevirt on ARM.

I believe Windows 10 requires UEFI boot on aarch64 anyway? I seem to recall the requirements were less lenient on that architecture.

Plus of course Windows 11 is now out, and it requires Secure Boot, which in turn requires UEFI, even on x86_64...

**Comment by hellocloudnative** [comment]
How to use sidecar-hook?   @zhlhahaha

**Comment by simonflood** [comment]
> Listening to the AllThingsOpen talk from @mazzystr and @stu-gott - showing off running kubevirt on a Raspberry Pi 4 w/8GB memory and 1T SSD. (I don't know when the talk recording will be published.)

https://youtu.be/RWhaC8LKP70

Description at https://2021.allthingsopen.org/sessions/running-vms-on-raspberry-pi-using-kubevirt/

**Comment by pcgeek86** [comment]
I'm having the same problem with the `a1.metal` virtual machine type on Amazon Web Services (AWS).

```
root@host# kubectl logs --namespace kubevirt pod/virt-operator-97778fbf8-69sdq
exec /usr/bin/virt-operator: exec format error
```

**Comment by rmohr** [comment]
@pcgeek86 can it be that you just used our release manifests? Arm is not part of our releases yet because it is not fully done (although close). You can install it like described here: https://kubevirt.io/user-guide/operations/installation/#experimental-arm64-developer-builds

**Comment by pcgeek86** [comment]
> @pcgeek86 can it be that you just used our release manifests for x86_64? Arm is not part of our releases yet because it is not fully done (although close). You can install it like described here: https://kubevirt.io/user-guide/operations/installation/#experimental-arm64-developer-builds

Thank you @rmohr! I will take a look at these experimental builds! I didn't see this documentation before. Cheers

**Comment by zhlhahaha** [comment]
> On https://quay.io/repository/kubevirt/virt-operator?tab=tags&tag=latest, the last ARM64 images are broken -- meaning that they are in fact for AMD64. I did a quick bisect, and the last image that worked was from around early February this year:
> 
> ```shell
> > docker run quay.io/kubevirt/virt-operator:20230212_5588ee3a9-arm64 -- /bin/bash
> WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
> exec /usr/bin/virt-operator: exec format error
> 
> > docker run quay.io/kubevirt/virt-operator:20230206_4834858d8-arm64 -- /bin/bash
> {"component":"virt-operator","level":"info","msg":"cannot find virt-operator's image\n","pos":"application.go:148","timestamp":"2023-05-03T15:34:27.130720Z"}
> ```

Thanks for inform, I will check it.

**Comment by zhlhahaha** [comment]
There are two issues, 1. cross-compile script is misconfigured. 2. cross compile tool chain is unable to download.

In terms of the second issue, there are some SSL certification issue when trying to download cross compile tool chain. I am contacting our IT team and trying to fix it. Once it fixed, then I can verify if my fix for the first issue works well. It may takes some time to make Arm64 images available.
Sorry about it. @ader1990

**Comment by andreabolognani** [comment]
@zhlhahaha interestingly, February is also when the aarch64 tags for `quay.io/kubevirt/fedora-with-test-tooling-container-disk` started pointing to the same hashes as the x86_64 ones. I wouldn't be surprised if both of these could be traced back to the same change to the container build process.

**Comment by zhlhahaha** [comment]
> @zhlhahaha interestingly, February is also when the aarch64 tags for `quay.io/kubevirt/fedora-with-test-tooling-container-disk` started pointing to the same hashes as the x86_64 ones. I wouldn't be surprised if both of these could be traced back to the same change to the container build process.

Yes, cross-build does not work from that moment.

**Comment by zhlhahaha** [comment]
> There are two issues, 1. cross-compile script is misconfigured. 2. cross compile tool chain is unable to download.


The issue 1 should be fixed by https://github.com/kubevirt/kubevirt/pull/9703
And the issue 2 has been fixed.

Hi @rmohr, shall we mirror the cross build tool chain in https://developer.arm.com/-/media/Files/downloads/gnu/11.3.rel1/binrel/arm-gnu-toolchain-11.3.rel1-x86_64-aarch64-none-linux-gnu.tar.xz
It is the second time that we are not able to download it because of SSL certificate issue.

**Comment by tarastella** [comment]
I've tried with the latest build (20230602) and the images with tag 20230602_e291f5f87-arm64 and 20230526_33dbfa922-arm64, but apparently I still have the "exec /bin/sh: exec format error"
Is there any specific tag that I can use as a test that is known to be working? Thanks.

**Comment by zhlhahaha** [comment]
> I've tried with the latest build (20230602) and the images with tag 20230602_e291f5f87-arm64 and 20230526_33dbfa922-arm64, but apparently I still have the "exec /bin/sh: exec format error" Is there any specific tag that I can use as a test that is known to be working? Thanks.

Hi @tarastella, the PR which fix the issue just got merged yestarday, https://github.com/kubevirt/kubevirt/pull/9703 . Would you give it a try for latest build. (20230603)

**Comment by tarastella** [comment]
Hi @zhlhahaha! Thanks for looking at this. Seems that 20230603 hasn't fixed things yet, I stil got the "exec format error". But it's understandable, as it might take another build or two.

Also, I am running a mixed hardware where the masters are amd64 and computes are arm64.
I understood that it's either arm64 or amd64. How do I force the operator to run on the arm nodes?

I already run this:
```
kubectl patch -n kubevirt kubevirt kubevirt --type merge --patch '{"spec": {"infra": {"nodePlacement": {"nodeSelector": {"compute": "true"}}}}}'
kubectl patch -n kubevirt kubevirt kubevirt --type merge --patch '{"spec": {"workloads": {"nodePlacement": {"nodeSelector": {"compute": "true"}}}}}'
```
Where "compute" is a label I applied to arm64 nodes.
Is anything else I shall modify?
Many thanks.

**Comment by jbpratt** [comment]
Hey @tarastella, I'm running the `20230603` on my single node arm64 cluster now but there are different manifests for arm64:

- https://storage.googleapis.com/kubevirt-prow/devel/nightly/release/kubevirt/kubevirt/20230603/kubevirt-operator-arm64.yaml
- https://storage.googleapis.com/kubevirt-prow/devel/nightly/release/kubevirt/kubevirt/20230603/kubevirt-cr-arm64.yaml

```
❯ kubectl get po -n kubevirt virt-controller-7c94d565c7-mzw7w -o yaml | grep image
    - --launcher-image
    - --exporter-image
    image: quay.io/kubevirt/virt-controller:20230603_0a3062212-arm64
    imagePullPolicy: IfNotPresent
    image: quay.io/kubevirt/virt-controller:20230603_0a3062212-arm64
    imageID: quay.io/kubevirt/virt-controller@sha256:0ace9e6226b357965be8b0c871264bdb79d8a0b3a0290081c405b4b94465dd4a

❯ kubectl logs -n kubevirt virt-controller-7c94d565c7-mzw7w
W0603 08:34:14.050823       1 client_config.go:618] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
{"component":"virt-controller","level":"info","msg":"STARTING informer kubeVirtInformer","pos":"virtinformers.go:345","timestamp":"2023-06-03T08:34:14.054292Z"}
{"component":"virt-controller","level":"info","msg":"STARTING informer CRDInformer","pos":"virtinformers.go:345","timestamp":"2023-06-03T08:34:14.054424Z"}
{"component":"virt-controller","level":"info","msg":"CDI detected, DataVolume integration enabled","pos":"application.go:395","timestamp":"2023-06-03T08:34:21.155944Z"}
...
```

**Comment by tarastella** [comment]
Weird behaviour. I've applied exactly those manifests on an existing kubevirt, but it hasn't applied the changes.
Anyway, I've deleted kubevirt and reinstalled again using those manifests. Plus, I had to patch infra/workload to run on the arm64 node.  With that, the control processes of kubevirt are running and the downloaded images are arm64 at the appropriate version.

I'm still stuck with the VM not running for some strange reason but I'm not understanding what it is.
I suspect I miss some kernel modules, as I'm running DietPI on a NanoPI R6S, and I can tell you I had to re-compile the kernel including some missing modules.

**Comment by tarastella** [comment]
@zhlhahaha yes, I made it at the end using a ceph backend (there is no much storage on the R6S left of the MMC).
I won't be able to work on it for a couple of weeks.
I need to do some more tests as it seems not to be 100% reliable, but I don't know if it's the master fault.

**Comment by dneary** [comment]
> > Looking forward to hearing progress on this, as [rancher/harvester#371](https://github.com/rancher/harvester/issues/371) (Harvester., a new hyperconverged infrastrucure from Rancher) depends on this working.
> 
> @vielmetti our main issue right now for supporting this is actually having ARM64 machines for building and CI. If rancher depends on it, maybe you can start a thread on the kubevirt-dev mailing list and see what we can do about this?

Is hardware availability still an issue?

**Comment by xpivarc** [comment]
> > > Looking forward to hearing progress on this, as [rancher/harvester#371](https://github.com/rancher/harvester/issues/371) (Harvester., a new hyperconverged infrastrucure from Rancher) depends on this working.
> > 
> > 
> > @vielmetti our main issue right now for supporting this is actually having ARM64 machines for building and CI. If rancher depends on it, maybe you can start a thread on the kubevirt-dev mailing list and see what we can do about this?
> 
> Is hardware availability still an issue?

@dneary No, the main issue is to enable full test coverage.

**Comment by senpro-ingwersenk** [comment]
@xpivarc Been a few months; whats the current progress?
I was planning to migrate my current setup of SBCs to a few RockChip-based mITX Servers and wanted to use harvester and am using k3s right now.
Thanks!

**Comment by zhlhahaha** [comment]
> senpro-ingwersenk

Hi, I did not try to integrate kubevirt to harvester on Arm64 platform, so I am not sure if it works well. You can give it a try.
And feel free to leave question here if you get any problem.

**Comment by senpro-ingwersenk** [comment]
Thanks @zhlhahaha !
I'll see if it works in "plain" kubernetes first (which is k3s in my case, which is also used by harvester afaik - so its a good standalone test). Thank you for the information :)

**Comment by xpivarc** [comment]
We are still missing full coverage, that being said I am pretty sure most things will work. Just setting the expectation that there are still monsters out there we yet need to meet.

**Comment by ader1990** [comment]
We have been running K8S on ARM64 for a while, and works as expected. All the upstream tooling supports now ARM64, including CAPI, KubeVirt, Rook/Ceph, ArgoCD and more. If you want to a quick proof of concept, I suggest the dockerized version of k3s called k3d, works great and does not taint the environment: https://k3d.io/v5.7.2/

Let me know if you have any questions.

**Comment by vasiliy-ul** [comment]
> I'll see if it works in "plain" kubernetes first (which is k3s in my case, which is also used by harvester afaik - so its a good standalone test). Thank you for the information :)

Harvester already provides Arm64 builds as tech preview (I think starting from v1.3.0): https://github.com/harvester/harvester/releases/tag/v1.3.1.

And btw, Harvester uses RKE2 instead of k3s. It has switched from k3s quite some time ago.

**Comment by nics90** [comment]
Still CDI doesn't work on arm and it's one of the blocker for production level use cases. Also, Microsoft recently released ARM supported Windows 11 Image, we tried that on Kubevirt and that also didn't work due to UEFI boot errors.

**Comment by ader1990** [comment]
> Still CDI doesn't work on arm and it's one of the blocker for production level use cases. Also, Microsoft recently released ARM supported Windows 11 Image, we tried that on Kubevirt and that also didn't work due to UEFI boot errors.

Hello @nics90, I did not have any issues with the CDI operator on ARM64 if used with the arm64 tagged docker images -> https://quay.io/repository/kubevirt/cdi-controller?tab=tags&tag=latest.

Related to the Windows on ARM64, I did try a while ago and did not work for me either.

**Comment by nics90** [comment]
Thanks @ader1990 : I tried using the above image but got below error:
kubectl logs cdi-operator-7cbc545b58-rm8g6 -n cdi
{"level":"info","ts":"2024-08-20T13:03:14Z","logger":"cmd","msg":"Go Version: go1.22.3 X:nocoverageredesign"}
{"level":"info","ts":"2024-08-20T13:03:14Z","logger":"cmd","msg":"Go OS/Arch: linux/arm64"}
{"level":"info","ts":"2024-08-20T13:03:14Z","logger":"cmd","msg":"Registering Components."}
{"level":"error","ts":"2024-08-20T13:03:14Z","logger":"cmd","msg":"","error":"required key OVIRT_POPULATOR_IMAGE missing value","stacktrace":"main.main\n\tcmd/cdi-operator/operator.go:147\nruntime.main\n\tGOROOT/src/runtime/proc.go:271"}

**Comment by ader1990** [comment]
> > Still CDI doesn't work on arm and it's one of the blocker for production level use cases. Also, Microsoft recently released ARM supported Windows 11 Image, we tried that on Kubevirt and that also didn't work due to UEFI boot errors.
> 
> Hello @nics90, I did not have any issues with the CDI operator on ARM64 if used with the arm64 tagged docker images -> https://quay.io/repository/kubevirt/cdi-controller?tab=tags&tag=latest.
> 
> Related to the Windows on ARM64, I did try a while ago and did not work for me either.

I was using `quay.io/kubevirt/cdi-controller:20240401_e262fb1d8-arm64` and do not have the issue.

```
kubectl --kubeconfig ~/kub-poc.kubeconfig logs deployment/cdi-operator -n cdi |grep -i 'register' -A 10 -B 10
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"cmd","msg":"Go Version: go1.21.5 X:nocoverageredesign"}
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"cmd","msg":"Go OS/Arch: linux/arm64"}
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"cmd","msg":"Registering Components."}
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"cdi-operator","msg":"","VARS":"{OperatorVersion:v1.59.0 ControllerImage:quay.io/kubevirt/cdi-controller:20240401_e262fb1d8-arm64 DeployClusterResources:true ImporterImage:quay.io/kubevirt/cdi-importer:20240401_e262fb1d8-arm64 ClonerImage:quay.io/kubevirt/cdi-cloner:20240401_e262fb1d8-arm64 APIServerImage:quay.io/kubevirt/cdi-apiserver:20240401_e262fb1d8-arm64 UploadProxyImage:quay.io/kubevirt/cdi-uploadproxy:20240401_e262fb1d8-arm64 UploadServerImage:quay.io/kubevirt/cdi-uploadserver:20240401_e262fb1d8-arm64 Verbosity:1 PullPolicy:IfNotPresent ImagePullSecrets:[] PriorityClassName: Namespace:cdi InfraNodePlacement:0x4000613470 ControllerReplicas:0 APIServerReplicas:0 UploadProxyReplicas:0}"}
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"cdi-operator","msg":"Unable to get controller reference, using namespace"}
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"cmd","msg":"Starting the Manager."}
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"controller-runtime.metrics","msg":"Starting metrics server"}
I0820 12:14:25.790801       1 leaderelection.go:250] attempting to acquire leader lease cdi/cdi-operator-leader-election-helper...
{"level":"info","ts":"2024-08-20T12:14:25Z","logger":"controller-runtime.metrics","msg":"Serving metrics server","bindAddress":":8080","secure":false}
I0820 12:14:25.802299       1 leaderelection.go:260] successfully acquired lease cdi/cdi-operator-leader-election-helper
{"level":"debug","ts":"2024-08-20T12:14:25Z","logger":"events","msg":"cdi-operator-64744ddd85-qwkjb_365886bc-2717-4cd9-826c-18cc91262099 became leader","type":"Normal","object":{"kind":"Lease","namespace":"cdi","name":"cdi-operator-leader-election-helper","uid":"8b9388c5-aa82-4aa3-b578-4c7ee738dade","apiVersion":"coordination.k8s.io/v1","resourceVersion":"26685"},"reason":"LeaderElection"}
{"level":"info","ts":"2024-08-20T12:14:25Z","msg":"Starting EventSource","controller":"cdi-operator-controller","source":"kind source: *v1beta1.CDI"}
{"level":"info","ts":"2024-08-20T12:14:25Z","msg":"Starting Controller","controller":"cdi-operator-controller"}

```
